repo2docker/buildpacks/julia/semver.py

```diff
@@ -0,0 +1,162 @@
+
+    comparison_symbol = constraint_str[0 : first_digit.start()].strip()
+
+    # Default to "^" search if no matching mode specified (up to next major version)
+    if (first_digit.start() == 0) or (comparison_symbol == "^"):
+        if major(constraint) == 0:
+            # Also, julia treats pre-1.0 releases specially, as if the first
+            # non-zero number is actually a major number:
+            # https://docs.julialang.org/en/latest/stdlib/Pkg/#Caret-specifiers-1
+            # So we need to handle it separately by bumping the first non-zero
+            # enumber.
+            for i, n in enumerate(constraint):
+                if (
+                    n != 0 or i == len(constraint) - 1
+                ):  # (using the last existing number handles situations like "^0.0" or "^0")
+                    upper = constraint[0:i] + (n + 1,)
+                    break
+            return VersionRange(constraint, upper, True)
+        else:
+            return VersionRange(
+                constraint, (major(constraint) + 1,), True
+            )
+
+    # '~' matching (only allowed to bump the last present number by one)
+    if comparison_symbol == "~":
+        return VersionRange(
+            constraint,
+            constraint[:-1] + (constraint[-1] + 1,),
+            exclusive=False
+        )
+
+    # Use semver package's comparisons for everything else:
+
+    # semver requires three version numbers
+    if len(constraint) < 3:
+        while len(constraint) < 3:
+            constraint = constraint + (0,)
+        constraint_str = constraint_str[0 : first_digit.start()] + ".".join(
+            map(str, constraint)
+        )
+
+    # Convert special comparison strings to format accepted by `semver` library.
+    constraint_str = constraint_str.replace("≥", ">=").replace("≤", "<=")
+    constraint_str = re.sub(r"(^|\b)=\b", "==", constraint_str)
+
+    return SemverMatcher(constraint_str)
+
+
+class SemverMatcher:
+    """Match a version tuple to a given `constraint_str`
+
+    The matching is handled via the `semver` package.
+    """
+
+    def __init__(self, constraint_str):
+        self.constraint_str = constraint_str
+
+    def match(self, v):
+        """Check if `v` matches the constraint"""
+        while len(v) < 3:
+            v = v + (0,)
+        v_str = ".".join(map(str, v))
+        return semver.match(v_str, self.constraint_str)
+
+    def __eq__(self, rhs):
+        return self.constraint_str == rhs.constraint_str
+
+    def __repr__(self):
+        return self.constraint_str
+
+
+class VersionRange:
+    """Test if a version tuple is between a lower and upper bound
+
+    A VersionRange represents a range of versions from `lower` to `upper`
+    and lets you test if a version falls in that range. The `upper` bound
+    can either be exclusive (exclusive=True) or inclusive (exclusive=False).
+    """
+
+    def __init__(self, lower, upper, exclusive):
```

Comment:
Let's set a sensible default for exclusive which I would recommend to be inclusive.